fiftyone/core/collections.py

```diff
@@ -9175,12 +9211,22 @@ def create_index(self, field_or_spec, unique=False, **kwargs):
             # Satisfactory index already exists
             return index_name
 
-        if is_frame_index:
-            coll = self._dataset._frame_collection
+        if using_async:
+            logging.info("Creating index '%s' asynchronously...", index_name)
+            coll = (
+                self._dataset._frame_collection_async
+                if is_frame_index
+                else self._dataset._sample_collection_async
+            )
+            name = await coll.create_index(index_spec, unique=unique, **kwargs)
```

Comment:
What happens if this is never `await`ed? Will the index still be created?

Reply:
Motor's version of `create_index` is a co-routine, so it must be awaited. 

Te wrapper should help with `dataset.create_index(...)` for both sync and async case

`dataset.create_index(...)` should function as normal, synchronously
`dataset.creata_index(..., using_async=True)` returns a co-routine, which should be awaited

Reply:
My question is slightly different: what happens if I do this:

```py
dataset.create_index(..., using_sync=True)
```

and don't await it? Will the index be created or not?

Reply:
The index won't be created as the coroutine is not executed

Reply:
Something like this:

```
In [4]: dataset = fo.load_dataset("dave-quickstart")
2024-08-07 15:43:55 - INFO - {"message": "Waiting for suitable server to become available", "selector": "Primary()", "operation": "buildinfo", "topologyDescription": "<TopologyDescription id: 66b3f8abf5462878245b98b9, topology_type: Unknown, servers: [<ServerDescription ('staging-mongodb.fiftyone.ai', 29017) server_type: Unknown, rtt: None>]>", "clientId": {"$oid": "66b3f8abf5462878245b98b9"}, "remainingTimeMS": 29}

In [5]: dataset.list_indexes()
Out[5]:
['id',
 'filepath',
 'ground_truth.detections.label',
 'ground_truth.detections.confidence',
 'ground_truth.detections.tags',
 'predictions.detections.tags',
 'predictions.detections.label',
 'ground_truth.detections.bounding_box',
 'ground_truth.detections.index',
 'ground_truth.detections.area']

In [6]: dataset.drop_index("ground_truth.detections.label")

In [7]: dataset.list_indexes()
Out[7]:
['id',
 'filepath',
 'ground_truth.detections.confidence',
 'ground_truth.detections.tags',
 'predictions.detections.tags',
 'predictions.detections.label',
 'ground_truth.detections.bounding_box',
 'ground_truth.detections.index',
 'ground_truth.detections.area']

In [8]: dataset.create_index("ground_truth.detections.label", using_async=True)
2024-08-07 15:46:29 - INFO - Running async, checking for running event loop
Out[8]: <coroutine object SampleCollection.create_index at 0x1037def10>

In [9]: dataset.list_indexes()
Out[9]:
['id',
 'filepath',
 'ground_truth.detections.confidence',
 'ground_truth.detections.tags',
 'predictions.detections.tags',
 'predictions.detections.label',
 'ground_truth.detections.bounding_box',
 'ground_truth.detections.index',
 'ground_truth.detections.area']

In [10]: dataset.list_indexes()
Out[10]:
['id',
 'filepath',
 'ground_truth.detections.confidence',
 'ground_truth.detections.tags',
 'predictions.detections.tags',
 'predictions.detections.label',
 'ground_truth.detections.bounding_box',
 'ground_truth.detections.index',
 'ground_truth.detections.area']

In [11]: dataset.create_index("ground_truth.detections.label")
2024-08-07 15:46:50 - INFO - Creating index 'ground_truth.detections.label' synchronously...
Out[11]: 'ground_truth.detections.label'

In [12]: dataset.list_indexes()
Out[12]:
['id',
 'filepath',
 'ground_truth.detections.confidence',
 'ground_truth.detections.tags',
 'predictions.detections.tags',
 'predictions.detections.label',
 'ground_truth.detections.bounding_box',
 'ground_truth.detections.index',
 'ground_truth.detections.area',
 'ground_truth.detections.label']
```

Reply:
Another solution is to keep `dataset.create_index(...)` as it is; I will refactor so that we have a completely different async method, `dataset.create_index_async(...)` which is identical, but in async mode instead :) In this way, we can avoid the complicated wrapper around `create_index(..., using_async=True)`.

Reply:
Okay my goal here is not to make `create_index()` async friendly per-se, but rather to see if there's a way for the code to complete (whether sync or async) and return while the index creation is still ongoing.

If there is a way, then we don't need to require [this code](https://github.com/voxel51/fiftyone-plugins/blob/2fe3f6bc59339f85ef199c432a97442dc040f12c/plugins/indexes/__init__.py#L38-L57) to be run via delegated execution; users could run the operator in immediate execution mode and their API request will complete although the index is still being built.

If you can't find a way for the index to be created without `await` or without synchronously waiting for `create_index()` to return, then we'll need to require users to run the operator I linked in delegated mode.

Make sense?

Reply:
Yes, we discussed this on Monday and I have been trying to find a way to send the command to build indexes via Mongo DB cluster but without success so far :) that's why I thought we can try utilizing the async Motor client since it is the official MongoDB async driver. I have been searching the documentation for RESTful CRUD APIs but it doesn't seem to be very popular; we can try to dive-deeper into Pymongo and try to pull up the URI and the command but I feel it was a little bit hacky.

Another approach would be to add a time to the sync `dataset.create_index(...)` so that it can terminate after a short period of time (a few seconds), but it might be tricky to time it since the client itself might take a while to open a connection to the DB cluster.

```
2024-08-07 13:04:14 - INFO - {"message": "Waiting for suitable server to become available", "selector": "Primary()", "operation": "buildinfo", "topologyDescription": "<TopologyDescription id: 66b3d33e902ba2daae4614da, topology_type: Unknown, servers: [<ServerDescription ('staging-mongodb.fiftyone.ai', 29017) server_type: Unknown, rtt: None>]>", "clientId": {"$oid": "66b3d33e902ba2daae4614da"}, "remainingTimeMS": 29}
```